### ISSUE TITLE
Running tests with `ginkgo` significantly slower than `go test`

### ISSUE THREAD

**Issue by charleskorn** [onsi/ginkgo#644]
Thanks for the awesome project!

I'm seeing an odd issue: running my test suites with `ginkgo` takes significantly longer than running them with `go test`:

```
$ time go test -tags unitTests ./server/...
ok  	github.com/batect/abacus/server/api	0.496s
ok  	github.com/batect/abacus/server/cmd	0.335s
ok  	github.com/batect/abacus/server/middleware	0.195s
?   	github.com/batect/abacus/server/middleware/testutils	[no test files]
ok  	github.com/batect/abacus/server/storage	0.173s
        2.69 real         5.60 user         1.93 sys
```

```
$ time ~/go/bin/ginkgo -tags unitTests server/...
[1581318849] API Suite - 46/46 specs •••••••••••••••••••••••••••••••••••••••••••••• SUCCESS! 2.535598ms PASS
[1581318849] Application Suite - 0/0 specs  SUCCESS! 241.13µs PASS
[1581318849] Middleware Suite - 9/9 specs ••••••••• SUCCESS! 778.471µs PASS
[1581318849] Storage Suite - 3/3 specs ••• SUCCESS! 206.337µs PASS

Ginkgo ran 4 suites in 4.485361489s
Test Suite Passed
        4.49 real         9.35 user         5.09 sys
```

I've tried setting `-compilers` to a low number as suggested in https://github.com/onsi/ginkgo/issues/258, but this doesn't have a significant impact. 

https://github.com/batect/abacus/commit/f8d4c577d4dc5fb39004d5cceb4f6384852d42df is the commit I'm testing this on.

**Comment by toddkazakov** [comment]
At Tidepool we're also seeing roughly 2 times slower test execution times. We have a very large test suite, which takes more than 8 minutes to complete with `ginkgo` vs 3.5 minutes with `go test`. Unfortunately, test randomization is only supported by `ginkgo`. Please advice how to profile this.

**Comment by elamre** [comment]
Also I notice a very slow performance of running ginkgo. Has any of you found a workaround?

**Comment by onsi** [comment]
sorry for the radio silence, y'all.  i'd like to get too the bottom of this.  `ginkgo` uses vanilla `go test` but it first *compiles* all the test suites it discovers  before running them.  my instinct is that the slowdown happens during the compilation phase and likely has something to do with the evolution of go in recent versions towards module support.

specifically `ginkgo` compiles the test suites it discovers in parallel.  so, a few requests:

- can anyone here point me at a public repo that reproduces this problem (please include go version, etc.)
- have you tried explicitly setting `-compilers=1` to ensure there is no concurrency *at all* during compilation?

we'll get to the bottom of this.  `ginkgo`'s just using `go test` after all so it should be about as fast.  2x slower makes no sense.

**Comment by onsi** [comment]
actually i see that @charleskorn is pointing to a repo that reproduces the issue.  i'll dig in and report back

**Comment by onsi** [comment]
I've made some progress outlined here:
https://github.com/onsi/ginkgo_perf

tl;dr: Ginkgo compiles test suites into a temp directory and then runs them.  I'm observing a latency of O(500ms) when launching compiled test suites stored in temp directories.  Changing Ginkgo's behavior to store compiled assets (temporarily) in the package path itself reduces this latency substantially.

Without this change, Ginkgo is O(500ms) slower than `go test -c -o out.test && out.test`.  With this change it is about as fast.

Note that Ginkgo is still O(100ms) slower than `go test` - but this is a limitation that will be hard to address as Ginkgo's approach to parallelism requires creating a compiled test artifact.

With the new behavior I see the ginkgo unit tests go from O(16.2s) down to O(12.7s).

I'm pushing a commit with the new behavior.  I don't expect it to address all the issues folks are experiencing, but it's a start.  Please give it a whirl.

@toddkazakov I'd love more detail on your usecase. Any chance there's a repo I can clone that reproduces the issue?

**Comment by onsi** [comment]
I'll note that I don't understand _why_ this change makes such a significant difference.  And it may well be OS/hardware dependent (I'm running MacOS on a macbook pro with an SSD drive).

**Comment by toddkazakov** [comment]
Thanks for your help @onsi. The issue can be reproduced in https://github.com/tidepool-org/platform. The tests can be executed with `make test`, however they require mongo to be running at the default port on localhost. 

The issue is more apparent in Travis due to the resource constraints. I tried removing all extra options (coverage, race condition detection, randomization, etc) and set the compilers to 1. There was no significant speed up. However, this still doesn't explain why plain `go test` is 4x faster.

Locally, the test compilation is parallelized and they run ok on a 6-core MBP with SSD.

**Comment by onsi** [comment]
thanks @toddkazakov i’ll take a look.  I assume you see the same discrepancy on your MPB between ginkgo and go test?  And I see that you don’t have `-p` or `-nodes` set in your makefile so I assume you are running the tests without ginkgo’s parallelism, correct?

**Comment by onsi** [comment]
hey @toddkazakov quick update:
- i'm seeing mongod crap out with `Abort trap: 6` when I run the test suite (and this, in turn, causes tests to hang/fail).  any thoughts on what i'm doing wrong?
- when i skip the mongo tests by passing `-skipPackage=mongo` to `ginkgo` I get the following run time:
  - With the previous version of the ginkgo CL (that uses tmp directories): 2m10s
  - With the new version (that doesn't use tmp directories): 1m46s

a difference of almost 30s!  Can you give the latest commit a try and see if you're also seeing improvements?

**Comment by toddkazakov** [comment]
The mongo error you're seeing is most likely caused by running a too recent version - we're using 3.6.

I ran a series of test today. Using the latest commit does provide a speed up locally, albeit small. I believe this will have a much bigger effect in Travis.
```
CI —compilers=1 / Latest commit
Ginkgo ran 155 suites in 5m3.807436627s
Test Suite Passed

real	5m3.867s
user	5m5.360s
sys	2m11.561s
```
vs the version we were using previously:
```
Ginkgo ran 155 suites in 5m17.166737852s
Test Suite Passed

real	5m17.236s
user	6m26.646s
sys	2m33.479s
```

I was able to make `go test` execution slower with `go test -p 1` and adding the extra options that we use currently (`-race` being the one that slows down the process the most). 

In my case I was comparing apples to oranges, but it was not obvious.

**Comment by onsi** [comment]
ah ok makes sense.  yes the race detector will slow things down substantially.

**Comment by onsi** [comment]
@elamre can you point me at a repo or give me a bit more detail so i can understand what sort of performance issues you are seeing?

**Comment by onsi** [comment]
Closing due to inactivit

**Comment by jerbob92** [comment]
Would it be an idea to limit this somehow?
If you have a computer with a lot of CPU cores, this can have dramatic effects on performance due to the amount of build processes it will start concurrently, this may also consume a lot of memory, which may cause the build process to become ever slower because of swapping.

I'm also wondering if it's able to utilize caching efficiently like this if it start a lot of build processes at the same time of tests that share a lot of packages.
